opengrok-tools/src/test/python/test_mirror.py

```diff
@@ -376,3 +398,183 @@ def mock_get_repo_type(*args):
             repos = get_repos_for_project(project_name, None, source_root,
                                           ignored_repos=['.'])
             assert len(repos) == 0
+
+
+@pytest.mark.parametrize(['config', 'expected_command'], [
+    (None, 'default-command'),
+    ('/usr/bin/git', '/usr/bin/git'),
+    ({}, 'default-command'),
+    ({'incoming': '/bin/false'}, 'default-command'),
+    ([], 'default-command'),
+    ({'command': '/usr/local/bin/git'}, '/usr/local/bin/git'),
+    (
+            {'command': '/usr/local/bin/git', 'incoming': '/bin/false'},
+            '/usr/local/bin/git'
+    )
+])
+def test_mirroring_custom_repository_command(config, expected_command):
+    assert expected_command == Repository._repository_command(
+        config,
+        lambda: 'default-command'
+    )
+
+
+@pytest.mark.parametrize(
+    ('touch_binary'), [
+        pytest.param('/bin/touch',
+                     marks=pytest.mark.skipif(
+                         not os.path.exists('/bin/touch'),
+                         reason="requires /bin binaries")),
+        pytest.param('/usr/bin/touch',
+                     marks=pytest.mark.skipif(
+                         not os.path.exists('/usr/bin/touch'),
+                         reason="requires /usr/bin binaries")),
+    ]
+)
+def test_mirroring_custom_incoming_invoke_command(touch_binary):
+    with tempfile.TemporaryDirectory() as repository_root:
+        repository = GitRepository(mock(), repository_root, 'test-1', {
+            'incoming': [touch_binary, 'incoming.txt']
+        }, None, None, None)
+        assert repository.incoming() is False
+        assert 'incoming.txt' in os.listdir(repository_root)
+
+
+@pytest.mark.parametrize(
+    ('echo_binary'), [
+        pytest.param('/bin/echo',
+                     marks=pytest.mark.skipif(
+                         not os.path.exists('/bin/echo'),
+                         reason="requires /bin binaries")),
+        pytest.param('/usr/bin/echo',
+                     marks=pytest.mark.skipif(
+                         not os.path.exists('/usr/bin/echo'),
+                         reason="requires /usr/bin binaries")),
+    ]
+)
+def test_mirroring_custom_incoming_changes(echo_binary):
+    with tempfile.TemporaryDirectory() as repository_root:
+        repository = GitRepository(mock(), repository_root, 'test-1', {
```

Comment:
do the tests assume that Git is installed on the system ?

Reply:
No no, the git command is never executed, most of the tests uses custom commands (like echo or true/false) and the rest uses mocking for the original call.